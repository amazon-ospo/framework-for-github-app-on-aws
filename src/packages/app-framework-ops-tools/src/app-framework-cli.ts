#!/usr/bin/env node

import { Command } from 'commander';
import { version } from '../package.json';
import { displayDynamoDBTables } from './getTableName';
import { importPrivateKey } from './importPrivateKey';

const program = new Command();
/**
 * CLI tool for Importing GitHub App private key into AWS KMS
 * Main command - 'app-framework'
 */
program
  .name('app-framework')
  .description(
    'CLI tool to get name of the App table with FrameworkForGitHubAppOnAwsManaged tag and to import GitHub App private key into AWS KMS',
  )
  .version(version);
// subcommand - getTableName
program
  .command('getTableName')
  .description('Displays App tables with FrameworkForGitHubAppOnAwsManaged tag')
  .action(async () => {
    try {
      await displayDynamoDBTables({});
    } catch (error) {
      console.error('Error:', error);
      process.exit(1);
    }
  });
// subcommand - importPrivateKey
program
  .command('importPrivateKey')
  .description('Import GitHub App private key into AWS KMS')
  .argument('<pemFilePath>', 'Path to the private key PEM file')
  .argument('<appId>', 'GitHub App ID')
  .argument('<tableName>', 'Table name to store the AppId and Key ARN')
  .addHelpText(
    'after',
    `
    Example:
      $ app-framework importPrivateKey private-key.pem 123456 my-table-name
  `,
  )
  .action(
    async (pemFilePath: string, appIdAsString: string, tableName: string) => {
      try {
        const appId = Number(appIdAsString);
        if (isNaN(appId)) {
          console.error('Error: GitHub AppId must be a valid number');
          process.exit(1);
        }

        await importPrivateKey({ pemFilePath, appId, tableName });
      } catch (error) {
        console.error('Error:', error);
        process.exit(1);
      }
    },
  );
program.parse(process.argv);
