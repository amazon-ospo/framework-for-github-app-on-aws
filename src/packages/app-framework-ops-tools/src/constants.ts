import { name, version } from '../package.json';
// Tags for DynamoDB App Table
export enum TagName {
  FRAMEWORK_FOR_GITHUB_APP_ON_AWS_MANAGED = 'FrameworkForGitHubAppOnAwsManaged',
  CREDENTIAL_MANAGER = 'CredentialManager',
  APP_TABLE = 'AppTable',
}

// Key Specifications
// Docs: https://docs.aws.amazon.com/kms/latest/APIReference/API_GetParametersForImport.html#API_GetParametersForImport_RequestSyntax
export const WRAPPING_SPEC = 'RSA_4096';
// Docs: https://docs.aws.amazon.com/kms/latest/APIReference/API_CreateKey.html#API_CreateKey_RequestSyntax
export const CREATE_KEY_SPEC = 'RSA_2048';

export const RESOURCES_PER_PAGE = 10;

export const USER_AGENT = `${name}/${version}`;
