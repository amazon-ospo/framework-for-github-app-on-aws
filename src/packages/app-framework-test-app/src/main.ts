import { CredentialManager } from '@aws/framework-for-github-app-on-aws';
import { App, Stack, StackProps, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';
// CDK App entry for @aws/framework-for-github-app-on-aws acceptance test.
// This stack is intended for testing @aws/framework-for-github-app-on-aws library.
// In a real use case, it should be a stack defined by customer.
export class TheAppFrameworkTestStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);
    const credentialManager = new CredentialManager(
      this,
      'CredentialManager',
      {},
    );
    const appTokenUrl = credentialManager.appTokenEndpoint;
    const installationAccessTokenUrl =
      credentialManager.installationAccessTokenEndpoint;
    new CfnOutput(this, 'AppTokenEndpoint', {
      value: appTokenUrl,
      exportName: 'AppTokenEndpoint',
    });
    new CfnOutput(this, 'InstallationAccessTokenEndpoint', {
      value: installationAccessTokenUrl,
      exportName: 'InstallationAccessTokenEndpoint',
    });
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new TheAppFrameworkTestStack(app, 'the-app-framework-test-stack', {
  env: devEnv,
});

app.synth();
