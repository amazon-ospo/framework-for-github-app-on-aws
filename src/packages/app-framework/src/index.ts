export * from './credential-manager/index';
