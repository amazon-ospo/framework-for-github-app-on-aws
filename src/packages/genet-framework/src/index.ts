export * from './credential-manager/index';
export * from './installation-manager/index';
