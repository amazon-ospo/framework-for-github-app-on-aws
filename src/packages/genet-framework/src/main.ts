import { App, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { CredentialManager } from './credential-manager';
import { InstallationManager } from './installation-manager';

// CDK App entry for Genet acceptance test.
export class Genet extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);
    new CredentialManager(this, 'CredentialManager', {});
    new InstallationManager(this, 'InstallationManager', {});
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new Genet(app, 'genet-framework', { env: devEnv });

app.synth();