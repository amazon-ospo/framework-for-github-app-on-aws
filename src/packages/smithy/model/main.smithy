$version: "2.0"

namespace framework.api

use aws.auth#sigv4
use aws.protocols#restJson1

@title("Framework for GitHub Apps on AWS")
@auth([sigv4])
@sigv4(name: "execute-api")
@restJson1
service AppFramework {
    version: "2024-08-23"
    resources: [
        CredentialManagementService
    ]
    errors: [
        AccessDeniedError,
        GatewayTimeoutError,
        ServiceUnavailableError
    ]
}

// Error that can occur when unable to access AWS resources
@error("client")
@httpError(403)
structure AccessDeniedError {
    message: String,
}

// Error that can occur when unable to access AWS resources
@httpError(504)
@error("server")
structure GatewayTimeoutError {
    message: String,
}

// Error that can occur when AWS service is unavailable
@httpError(503)
@error("server")
structure ServiceUnavailableError {
    message: String,
}

// Error that can occur when not authorized to access AWS resources
@httpError(401)
@error("client")
structure NotAuthorizedError {
    message: String,
}