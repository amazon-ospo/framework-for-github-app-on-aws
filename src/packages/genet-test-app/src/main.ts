import { App, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { CredentialManager, InstallationManager } from 'genet-framework';

// CDK App entry for Genet acceptance test.
// This stack is intended for testing genet-framework library.
// In a real use case, it should be a stack defined by customer.
export class GenetFrameworkTestStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);
    new CredentialManager(this, 'CredentialManager', {});
    new InstallationManager(this, 'InstallationManager', {});
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new GenetFrameworkTestStack(app, 'genet-framework-test-stack', { env: devEnv });

app.synth();
