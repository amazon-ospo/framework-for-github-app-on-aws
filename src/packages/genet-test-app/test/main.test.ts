import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { GenetFrameworkTestStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new GenetFrameworkTestStack(
    app,
    'genet-framework-test-stack',
    {},
  );

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
