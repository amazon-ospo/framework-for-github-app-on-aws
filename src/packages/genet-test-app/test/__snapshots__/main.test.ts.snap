// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CredentialManagerNestedStackCredentialManagerNestedStackResource341E8524": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Tags": [
          {
            "Key": "GenetComponent",
            "Value": "CredentialManager",
          },
        ],
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region",
              },
              ".",
              {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/ad713d4e978f839cee6211a8140e4fdc722fb58b487565a266e3f29573f8e052.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "InstallationManagerNestedStackInstallationManagerNestedStackResourceBB60C6B1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Tags": [
          {
            "Key": "GenetComponent",
            "Value": "InstallationManager",
          },
        ],
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region",
              },
              ".",
              {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/e9aa1c2a22ef3b492f1a21103200cecf7a84aee7235149d536ff4df4aa4fdd3f.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
